### PR TITLE
feature: adding zendesk destination connector

### PR DESCRIPTION
**Pull request by praneeth-ais**
## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] New Connector (Destination or Source Connector)
- [ ] Breaking change (fix or feature that would impact existing functionality)
- [ ] Styling change
- [ ] Documentation update
- [ ] Refactoring
- [ ] Chore

## Checklist:

- [x] Ensure a branch name is prefixed with `feature`, `bugfix`, `hotfix`, `release`, `style` or `chore` followed by `/` and branch name e.g `feature/add-salesforce-connector`
- [x] Added unit tests for the changes made (if required)
- [ ] Have you made sure the commit messages meets the guidelines?
- [x] Added relevant screenshots for the changes
- [x] Have you tested the changes on local/staging?
- [x] Added the new connector in rollout.rb
- [x] Have you updated the version number of the gem?
- [ ] Have you ensured that your changes for new connector are documented in the [docs repo](https://github.com/Multiwoven/docs) or relevant documentation files?
- [x] Have you updated the run time dependency in multiwoven-integrations.gemspec if new gems are added
- [x] Have you made sure the code you have written follows the best practises to the best of your knowledge?

### CODE REVIEW

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
+module Multiwoven
+    module Integrations
+        module Destination
+            module Zendesk
+                include Multiwoven::Integrations::Core
+
+                API_VERSION = "59.0"
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:7]: API_VERSION is not used

File: integrations/lib/multiwoven/integrations/core/constants.rb [hunk 2]
```diff
@@ -26,6 +26,8 @@ module Constants
         }
       }.freeze
 
+      ZENDESK_TICKETING_URL = "https://multiwoven-test.zendesk.com/api/v2/"
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/core/constants.rb:29]: is ZENDESK_TICKETING_URL constant across all customers?

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                    def initialize_client(connection_config)
+                        connection_config = connection_config.with_indifferent_access
+                        @client = ZendeskAPI::Client.new do |config|
+                            config.url = ZENDESK_TICKETING_URL
+                            config.username = connection_config[:username]
+                            config.password = connection_config[:password]
+                        end
+                    end
+
+                    def authenticate_client
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:49]: Is there an alternative method to check the connection status besides fetching tickets?

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+                    def discover(_connection_config = nil)
+                        catalog = build_catalog(load_catalog)
+                        catalog.to_multiwoven_message
+                        rescue StandardError => e
+                            handle_exception("ZENDESK:DISCOVER:EXCEPTION", "error", e)
+                            failure_status(e)
+                    end
+
+                    def write(sync_config, tickets, action = "create")
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:29]: nitpick: rename tickets to records. All the write method signature we are using records

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 3 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:49]: There are but this seemed like the most straightforward, simplest method.

I could be wrong though. Would you be willing to take a look at the API reference and confirm that I am not missing anything?

File: integrations/lib/multiwoven/integrations/core/constants.rb (hunk 2 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/core/constants.rb:29]: No. It depends on each customer's subdomain. I created the multiwoven-test subdomain for my tests but I will make this dynamic so it reads something like "https://#{subdomain}.zendesk.com/api/v2".

Thanks for calling this out.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 1 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:7]: I will delete it.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 4 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb]: This is done.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 1 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb]: Deleted.

File: integrations/lib/multiwoven/integrations/destination/zendesk/config/spec.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    "properties": {
+      "username": {
+        "type": "string",
+        "title": "Username",
+        "order": 0
+      },
+      "password": {
+        "type": "string",
+        "title": "Password",
+        "order": 1
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/zendesk/config/spec.json:21]: Please add `multiwoven_secret: true` for masking the password in frontend

File: integrations/lib/multiwoven/integrations/core/constants.rb (hunk 2 above)
Comment (karthik-sivadas) [reply integrations/lib/multiwoven/integrations/core/constants.rb:29]: Also please add this configuration in `spec.json` as this will change for every user

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+              config.url = ZENDESK_TICKETING_URL
+              config.username = connection_config[:username]
+              config.password = connection_config[:password]
+            end
+          end
+
+          def authenticate_client
+            response = @client.tickets.page(1).per_page(1).fetch
+
+            raise StandardError, "Failed to authenticate with Zendesk: #{response.errors.join(", ")}" unless response.is_a?(Array) && !response.empty?
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:51]: why are we raising an error if the response is empty and not an error? How is this related to authentication? What if the user has zero tickets then will it raise an error?

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+            true
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:53]: Please refrain from using true as a result for the function as this is a bad practice

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+          end
+
+          private
+
+          def initialize_client(connection_config)
+            connection_config = connection_config.with_indifferent_access
+            @client = ZendeskAPI::Client.new do |config|
+              config.url = ZENDESK_TICKETING_URL
+              config.username = connection_config[:username]
+              config.password = connection_config[:password]
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:44]: Ideally, it's advisable to use api token instead of password for security and maintainability reason especially when used in a third party application

File: integrations/lib/multiwoven/integrations/destination/zendesk/config/spec.json (hunk 5 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/config/spec.json:21]: done. committed.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 6 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:51]: It's related to authentication because it validates that there is a connection to a Zendesk account. The response body should contain a list/array of tickets. The list can be empty (user has zero tickets) so there will be no error raised. However, if the response itself is nil (response is empty), then an error will be raised.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 7 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:53]: removed this.

File: integrations/lib/multiwoven/integrations/core/constants.rb (hunk 2 above)
Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/core/constants.rb]: Done @karthik-sivadas.

File: integrations/.github/pull_request_template.md [hunk 9]
```diff
@@ -7,20 +7,20 @@
  Relates to issue #123 - 'Improve Salesforce Connector Efficiency'. If none, state 'None'. -->
 
 ## Type of Change
-- [ ] New Connector (Destination or Source Connector)
+- [x] New Connector (Destination or Source Connector)
```
Comment (karthik-sivadas) [inline integrations/.github/pull_request_template.md:10]: @praneeth-ais We are not suppose to update the template with the changes we need to update the description of the PR with these changes please revert this change

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb (hunk 6 above)
Comment (karthik-sivadas) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:51]: @praneeth-ais The purpose of this function is to validate the authentication if we get any other response then a 403 then it is authenticated even if it's null then why do we need this.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            catalog.to_multiwoven_message
+          rescue StandardError => e
+            handle_exception("ZENDESK:DISCOVER:EXCEPTION", "error", e)
+            failure_status(e)
+          end
+
+          def write(sync_config, records, action = "create")
+            @action = sync_config.stream.action || action
+            initialize_client(sync_config.destination.connection_specification)
+            process_records(records, sync_config.stream)
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:31]: write method should return tracking_message(write_success, write_failure)

Comment (praneeth-ais) [reply integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:31]: added.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                name: record[:name],
+                email: record[:email],
+                role: record[:role]
+              }
+            else
+              raise StandardError, "Unsupported record type: #{type}"
+            end
+          end
+
+          def success_status
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:105]: we can re-use this with the already defined method in the base connector.

File: integrations/lib/multiwoven/integrations/destination/zendesk/client.rb [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            ConnectionStatus.new(status: ConnectionStatusType["succeeded"]).to_multiwoven_message
+          end
+
+          def failure_status(error)
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/client.rb:109]: same here

File: integrations/lib/multiwoven/integrations/destination/zendesk/config/catalog.json [hunk 13]
```diff
@@ -0,0 +1,110 @@
+{
+    "request_rate_limit": 700,
+    "request_rate_limit_unit": "minute",
+    "request_rate_concurrency": 10,
+    "streams": [
+        {
+            "name": "Tickets",
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/zendesk/config/catalog.json:7]: We usually use the same name of the API path i.e `tickets` in https://developer.zendesk.com/api-reference/ticketing/introduction/ or function name in the gem which is also `tickets` and this is sent to the function inside the client via frontend. So we can avoid using pluralization etc

File: integrations/lib/multiwoven/integrations/destination/zendesk/icon.svg
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/zendesk/icon.svg:1]: why icon is not showng?

**Review by afthabvp — APPROVED** [review]
@praneeth-ais LGTM. Could you please fix the icon issue?

**Comment by afthabvp** [comment]
@karthik-sivadas LGTM. Please do a final review and merge this PR also in the Multiwoven-EE repo.

**Review by afthabvp — APPROVED** [review]

**Review by afthabvp — APPROVED** [review]

**Review by karthik-sivadas — APPROVED** [review]
